clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java

```diff
@@ -247,17 +138,32 @@ public boolean completeExceptionally(Throwable newException) {
      */
     @Override
     public synchronized boolean cancel(boolean mayInterruptIfRunning) {
-        return completeExceptionally(new CancellationException()) || exception instanceof CancellationException;
+        return completableFuture.cancel(mayInterruptIfRunning);
+    }
+
+    // We need to deal with differences between KF's historic API and the API of CF:
+    // CF#get() does not wrap CancellationException in ExecutionException (nor does KF).
+    // CF#get() always wraps the _cause_ of a CompletionException in ExecutionException (which KF does not).
+    //
+    // The semantics for KafkaFuture are that all exceptional completions of the future (via #completeExceptionally() or exceptions from dependants)
+    // manifest as ExecutionException, as observed via both get() and getNow().
+    private void maybeRewrapAndThrow(Throwable cause) {
+        if (cause instanceof CancellationException) {
+            throw (CancellationException) cause;
+        }
     }
 
     /**
      * Waits if necessary for this future to complete, and then returns its result.
      */
     @Override
     public T get() throws InterruptedException, ExecutionException {
-        SingleWaiter<T> waiter = new SingleWaiter<>();
-        addWaiter(waiter);
-        return waiter.await();
+        try {
+            return completableFuture.get();
+        } catch (ExecutionException e) {
+            maybeRewrapAndThrow(e.getCause());
+            throw e;
```

Comment:
nit: Would it make sense to move `throw e` into `maybeRewrapAndThrow` to let `maybeRewrapAndThrow` throw in both cases? More generally, I wonder if we could handle all the case in `maybeRewrapAndThrow` and use it everywhere.

Reply:
I agree it's a bit messy, but I don't think that works. The call site in `getNow()` differs from the others:

1. It's catching `CompletionException`, not `ExecutionException`
2. In the case there `maybeRewrapAndThrow` didn't throw it needs to wrap the cause in an `ExceptionException`, unlike the other call sites, to conform with the historical API of KafkaFuture. 

The first would require `maybeRewrapAndThrow()` to have an `Exception`-typed parameter, which, being a checked exception, means we'd need either to declare `throws Exception`, or we'd need to type cast within `maybeRewrapAndThrow()` and be left with an `else` case to deal with. 

The second is difficult to abstract over nicely for the other call sites: We'd need to pass a boolean to indicate whether the `ExceptionException` was needed.

Reply:
How about defining two helper methods, one for each cases?
* `private void maybeRewrapAndThrow(ExecutionException exception)`; and
* `private void maybeRewrapAndThrow(CompletionException exception)`

Reply:
@tombentley What do you think about this? Would it help to consolidate the logic?

Reply:
It doesn't seem to help with other places where we catch and rethrow because they catch unchecked exceptions `Error` and `RuntimeException` whose only supertype is `Throwable`, so we end up in a painful place due to Java's lack of abstraction for "unchecked exception".

For the existing call sites the `void` return type you proposed doesn't work with javac's definite return analysis so I have to declare `private <T> T rewrapAndThrow(CompletionException exception)` and `private <T> T rewrapAndThrow(ExecutionException exception)`. The method always throws, so the return type of `T` is just there to appease `javac` via a `return rewrapAndThrow(...)` at the call site. 

I can do this is you want, but I feel it's making the code less clear. The `rewrapAndThrow(CompletionException exception)` only has a single caller, so it's of little benefit.

I guess the other alternative would be to inline the method and just comment each site as to what's going on.

Reply:
I don't like the fact that we throw in two places... Could we at least make the name of `maybeRewrapAndThrow` a bit more explicit? It is only about `CancellationException` in the end so we could name it `maybeThrowCancellationException` or something like this. Moreover, the method does not really "rewrap" anything, right? It just checks the type and throw it.

Reply:
There don't seem to be any good options here.

We could throw from a single method if you're willing to have the method like this

```java
    private <T, E extends Exception> T throwUnwrappingCancellationException(E exception) throws E {
        if (exception.getCause() instanceof CancellationException) {
            throw (CancellationException) exception.getCause();
        }
        throw exception;
    }
```

and the call sites to look like this:

```java
return throwUnwrappingCancellationException(e);
```

 The `return <methodThatAlwaysThrows>` is nasty, and I can't think of a good method name. Personally I don't think this is better than naming the current method `maybeThrowCancellationException` as you suggest, and living with the throw in two places, so I've gone with that.